rust/actyx/ax-aql/src/language/aql.pest

```diff
@@ -7,23 +7,68 @@ main_timestamp = @{ SOI ~ isodate ~ EOI }
 
 // pragmas
 pragma = ${ "PRAGMA" ~ WHITE_SPACE+ ~ feature_word ~ ( !NEWLINE ~ WHITE_SPACE )* ~ ( pragma_multi | pragma_single ) }
-pragma_multi = _{ NEWLINE ~ pragma_multi_text ~ NEWLINE ~ "ENDPRAGMA" ~ NEWLINE }
-pragma_multi_text = ${ ( !( NEWLINE ~ "ENDPRAGMA" ~ NEWLINE ) ~ ANY )* }
+pragma_multi = _{ NEWLINE ~ pragma_multi_text ~ NEWLINE ~ pragma_end ~ NEWLINE }
+pragma_multi_text = ${ ( !( NEWLINE ~ pragma_end ~ NEWLINE ) ~ ANY )* }
+pragma_end = { "ENDPRAGMA" }
 pragma_single = _{ ":=" ~ ( !NEWLINE ~ WHITE_SPACE )* ~ pragma_single_text ~ NEWLINE }
 pragma_single_text = ${ ( !NEWLINE ~ ANY )* }
 
+features = { "FEATURES(" ~ feature_word* ~ ")" }
+feature_word = @{ ( ASCII_ALPHANUMERIC | "ø" )+ }
+
+// types
+type = { type_atom ~ type_post* ~ ( type_infix ~ type_atom ~ type_post* )* }
```

Comment:
is type_post here for describing something like `number[]`?
In this case, I find it foreign that `type_post` contains `type_dict` which allows `number{}`.
What is this for?

Reply:
Array is a function from [0, N) ↦ T, tuple does the same but the returned type depends on the index.
Record corresponds to a tuple, but with arbitrary names instead of the first N natural numbers as index.
Dict(ionary) is the corresponding feature to array, but again indexed by name; this is what is called `Record<T, U>` in Typescript, but with T fixed to be `BOOL | NUMBER | STRING`. For symmetry with the array notation, I chose to suffix with the same kind of parentheses used for a record.

```
tuple: [T1, T2, T3]
array: T[]

record: { n1: T1, n2: T2 }
dict:   T{}
```

Reply:
Ah I see now